src/login/logind-session-device.c

```diff
@@ -405,21 +406,20 @@ int session_device_new(Session *s, dev_t dev, bool open_device, SessionDevice **
 void session_device_free(SessionDevice *sd) {
         assert(sd);
 
-        /* Make sure to remove the pushed fd. */
+        /* Make sure to remove the pushed fd.
+         * Note DRM fds can hold a lot of memory :). */
         if (sd->pushed_fd) {
-                _cleanup_free_ char *m = NULL;
+                char m[sizeof("FDSTOREREMOVE=1\n" "FDNAME=\n") + 255];
                 const char *id;
-                int r;
 
                 /* Session ID does not contain separators. */
                 id = sd->session->id;
                 assert(*(id + strcspn(id, "-\n")) == '\0');
 
-                r = asprintf(&m, "FDSTOREREMOVE=1\n"
-                                 "FDNAME=session-%s-device-%u-%u\n",
-                                 id, major(sd->dev), minor(sd->dev));
-                if (r >= 0)
-                        (void) sd_notify(false, m);
+                xsprintf(m, "FDSTOREREMOVE=1\n"
+                            "FDNAME=session-%s-device-%u-%u\n",
+                            id, major(sd->dev), minor(sd->dev));
+                (void) sd_notify(false, m);
```

Comment:
Thanks for looking at the update.  I decided to avoid the use of `sd_notifyf()` for any FDSTOREREMOVE request.  Does this make it clear?

> It now reduces some anxiety I had about sd_notifyf() or equivalent returning -ENOMEM... when we're trying to free memory. The amended version uses stack allocation instead. The justification [included in the commit message] is still a bit messy.